src/Servant/Utils/Links.hs

```diff
@@ -6,77 +7,135 @@
+-- will result in a type error like this:
 --
--- __N.B.:__ 'mkLink' assumes a capture matches any string (without slashes).
+-- >>> let bad_link = Proxy :: Proxy ("hello" :> Delete)
+-- >>> safeLink bad_link  api
+-- <BLANKLINE>
+-- <interactive>:56:1:
+--     Could not deduce (Or
+--                         (IsElem' Delete (Get Int))
+--                         (IsElem'
+--                            ("hello" :> Delete)
+--                            ("bye" :> (QueryParam "name" String :> Delete))))
+--       arising from a use of ‘safeLink’
+--     In the expression: safeLink bad_link api
+--     In an equation for ‘it’: it = safeLink bad_link api
+--
+--  This error is essentially saying that the type family couldn't find
+--  bad_link under api after trying the open (but empty) type family
+--  `IsElem'` as a last resort.
 module Servant.Utils.Links (
-    -- * Link and mkLink
-    -- | The only end-user utilities
-      mkLink
-    , Link
-    -- * Internal
-    -- | These functions will likely only be of interest if you are writing
-    -- more API combinators and would like to extend the behavior of
-    -- 'mkLink'
-    , ValidLinkIn()
-    , VLinkHelper(..)
-    , IsElem
-    , IsLink
-                           )where
+  -- * Building and using safe links
+  --
+  -- | Note that 'URI' is Network.URI.URI from the network-uri package.
+    safeLink
+  , URI(..)
+  -- * Adding custom types
+  , HasLink(..)
+  , linkURI
+  , Link
+  , IsElem'
+  -- * Illustrative exports
+  , IsElem
+  , Or
+) where
 
+import Data.List
 import Data.Proxy ( Proxy(..) )
-import GHC.TypeLits ( KnownSymbol, Symbol, symbolVal )
+import Data.Text (Text, unpack)
+import Data.Monoid ( Monoid(..), (<>) )
+import Network.URI ( URI(..), escapeURIString, isUnreserved )
+import GHC.TypeLits ( KnownSymbol, symbolVal )
+import GHC.Exts(Constraint)
 
+import Servant.Common.Text
 import Servant.API.Capture ( Capture )
 import Servant.API.ReqBody ( ReqBody )
 import Servant.API.QueryParam ( QueryParam, QueryParams, QueryFlag )
 import Servant.API.MatrixParam ( MatrixParam, MatrixParams, MatrixFlag )
+import Servant.API.Header ( Header )
 import Servant.API.Get ( Get )
 import Servant.API.Post ( Post )
 import Servant.API.Put ( Put )
 import Servant.API.Delete ( Delete )
 import Servant.API.Sub ( type (:>) )
+import Servant.API.Raw ( Raw )
 import Servant.API.Alternative ( type (:<|>) )
 
+-- | If either a or b produce an empty constraint, produce an empty constraint.
+type family Or (a :: Constraint) (b :: Constraint) :: Constraint where
+    Or () b       = ()
+    Or a ()       = ()
 
-type family Or a b where
-    Or 'False 'False = 'False
-    Or 'True b       = 'True
-    Or a 'True       = 'True
-
-type family And a b where
-    And 'True 'True = 'True
-    And a 'False    = 'False
-    And 'False b    = 'False
+-- | You may use this type family to tell the type checker that your custom type
+-- may be skipped as part of a link. This is useful for things like
```

Comment:
Yes, I'm sure you could do all kinds of crazy things. I can't think of a useful thing to do other than skip, though.